spec/support/shared_examples/has_discovery_engine_control.rb

```diff
@@ -0,0 +1,115 @@
+RSpec.shared_examples "a model that has a Discovery Engine control" do
+  let(:control) do
+    instance_double(DiscoveryEngine::Control, create!: true, update!: true, delete!: true)
+  end
+
+  before do
+    allow(DiscoveryEngine::Control).to receive(:new).with(record).and_return(control)
+  end
+
+  describe "#save_with_control for a new record" do
+    subject(:record) { described_class.new(attributes) }
+
+    it "creates the control" do
+      record.save_with_control
+
+      expect(control).to have_received(:create!)
+    end
+
+    it "persists the record" do
+      record.save_with_control
+
+      expect(record).to be_persisted
+    end
+
+    context "when the control creation fails" do
+      let(:error) { DiscoveryEngine::Error.new("Uh oh") }
+
+      before do
+        allow(control).to receive(:create!).and_raise(error)
+      end
+
+      it "does not persist the record" do
+        record.save_with_control
+
+        expect(record).not_to be_persisted
+      end
+
+      it "adds an error to the record" do
+        record.save_with_control
+
+        expect(record.errors).to be_of_kind(:base, :discovery_engine_unrecoverable)
+      end
+    end
+  end
+
+  describe "#save_with_control for an existing record" do
+    subject(:record) { described_class.create!(attributes) }
+
+    before do
+      record.name = "New name"
+    end
+
+    it "updates the control" do
+      expect(control).to receive(:update!)
+
+      record.save_with_control
+    end
+
+    context "when the control update fails" do
+      let(:error) { DiscoveryEngine::Error.new("Uh oh") }
+
+      before do
+        allow(control).to receive(:update!).and_raise(error)
+      end
+
+      it "does not persist the changes to the record" do
+        record.save_with_control
+
+        expect(record).to be_changed
```

Comment:
Should this be `not_to` rather than `to`?

Reply:
Agree this doesn't read amazingly, but `changed?` refers to _unsaved_ changes, so we're expecting it to be true because the record has failed to save (and thus is still "changed").